paramiko/pipe.py

```diff
@@ -46,10 +46,9 @@ def __init__(self):
         self._closed = False
 
     def close(self):
+        self._closed = True
         os.close(self._rfd)
         os.close(self._wfd)
```

Comment:
@pettitpeon, is there any chance of either of these two `.close()` calls failing, making `self._closed == True` a possibly inaccurate representation of the pipe state?

(I would think that a failure of either of these two calls would signal an error state for the pipe such that `self._closed == True` is a _better_ state than `== False`... but I thought I'd raise the question.)

Reply:
> is there any chance of either of these two .close() calls failing?

Yes, it could happen. If either `.close()` fails, it would raise an `OSError`. In that case some FD might stay open, but the Pipe is not usable anyways. There is no use-case of using it after `.close()`. Worst case scenario, we leak resources until the application exits. The most probable scenario is that the close() exception is not handled and your program terminates anyways.  
**A small improvement** would be to have single `try`s on each `.close()`. In that case if, the first fails and leaks, we might still close the second one correctly.  
In any case, I see `.close()` errors as non-handable, we close() file descriptors as a "best-effort" to release the resource, but cannot really do anything if it fails.

>  making `self._closed == True` a possibly inaccurate representation of the pipe state?

Not really, after `close()` has been called, the object is disposed and no one should it anyways. The "broken state" is meaningless.

Last, the `pipe` is most probably a child of `Channel` and when the channel gets destroyed it tries to re-close, which  fails if the pipe was previously closed, but it is disregarded by a catch-all try. This is OK, and aligns with the idea that it is a non-handable error. https://github.com/paramiko/paramiko/blob/f4d7fecc1e636248db0adc40f136c1a77220b9da/paramiko/channel.py#L136